xds/internal/balancer/clusterresolver/balancerconfigbuilder/configbuilder.go

```diff
@@ -0,0 +1,308 @@
+			ChildPolicy: &internalserviceconfig.BalancerConfig{
+				Name:   weightedtarget.Name,
+				Config: wtConfig,
+			},
+			LoadReportingServerName: lrsServer,
+			MaxConcurrentRequests:   maxRequest,
+			DropCategories:          drops,
+		}
+		retAddrs = append(retAddrs, addrs...)
+	}
+
+	return retNames, retConfigs, retAddrs
+}
+
+// groupLocalitiesByPriority returns the localities grouped by locality.
+//
+// It also returns a list of strings where each string represents a priority,
+// and the list is sorted from higher priority to lower priority.
+func groupLocalitiesByPriority(localities []xdsclient.Locality) ([]string, map[string][]xdsclient.Locality) {
+	var priorityIntSlice []int
+	priorities := make(map[string][]xdsclient.Locality)
+	for _, locality := range localities {
+		if locality.Weight == 0 {
+			continue
+		}
+		priorityName := fmt.Sprintf("%v", locality.Priority)
+		priorities[priorityName] = append(priorities[priorityName], locality)
+		priorityIntSlice = append(priorityIntSlice, int(locality.Priority))
+	}
+	// Sort the priorities based on the int value, deduplicate, and then turn
+	// the sorted list into a string list. This will be child names, in priority
+	// order.
+	sort.Ints(priorityIntSlice)
+	priorityIntSliceDeduped := dedupSortedIntSlice(priorityIntSlice)
+	priorityNameSlice := make([]string, 0, len(priorityIntSliceDeduped))
+	for _, p := range priorityIntSliceDeduped {
+		priorityNameSlice = append(priorityNameSlice, fmt.Sprintf("%v", p))
+	}
+	return priorityNameSlice, priorities
+}
+
+func dedupSortedIntSlice(a []int) []int {
+	if len(a) == 0 {
+		return a
+	}
+	i, j := 0, 1
+	for ; j < len(a); j++ {
+		if a[i] == a[j] {
+			continue
+		}
+		i++
+		if i != j {
+			a[i] = a[j]
+		}
+	}
+	return a[:i+1]
+}
+
+// localitiesToWeightedTarget takes a list of localities (with the same
+// priority), and generates a weighted target config, and list of addresses.
+//
+// The addresses have path hierarchy set to [p, locality-name], so priority and
+// weighted target know which child policy they are for.
+func localitiesToWeightedTarget(localities []xdsclient.Locality, priorityName string, childPolicy *internalserviceconfig.BalancerConfig, lrsServer *string, cluster, edsService string) (*weightedtarget.LBConfig, []resolver.Address) {
+	weightedTargets := make(map[string]weightedtarget.Target)
+	var addrs []resolver.Address
+	for _, locality := range localities {
+		localityStr, err := locality.ID.ToString()
+		if err != nil {
+			localityStr = fmt.Sprintf("%+v", locality.ID)
+		}
+
+		var child *internalserviceconfig.BalancerConfig
+		if lrsServer != nil {
+			localityID := locality.ID
+			child = &internalserviceconfig.BalancerConfig{
+				Name: lrs.Name,
+				Config: &lrs.LBConfig{
+					ClusterName:             cluster,
+					EDSServiceName:          edsService,
+					ChildPolicy:             childPolicy,
+					LoadReportingServerName: *lrsServer,
+					Locality:                &localityID,
+				},
```

Comment:
If we make one at the top, we need to make copies of it in the loop, but it's a little hard because of the pointers.

And we still need to check if setting locality ID is necessary.

---

I can make it a lambda function though...

```go
	childCF := func(*internal.LocalityID) *internalserviceconfig.BalancerConfig {
		return childPolicy
	}
	if lrsServer != nil {
		childCF = func(lid *internal.LocalityID) *internalserviceconfig.BalancerConfig {
			return &internalserviceconfig.BalancerConfig{
				Name: lrs.Name,
				Config: &lrs.LBConfig{
					ClusterName:             cluster,
					EDSServiceName:          edsService,
					ChildPolicy:             childPolicy,
					LoadReportingServerName: *lrsServer,
					Locality:                lid,
				},
			}
		}
	}

    for _, locality := range localities {
        ...

		weightedTargets[localityStr] = weightedtarget.Target{
			Weight:      locality.Weight,
			ChildPolicy: childCF(&locality.ID),
		}
    }
```